File: backend/layers/common/nodejs/src/dynamo-client.ts
import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient } from '@aws-sdk/lib-dynamodb';

const client = new DynamoDBClient({});

export const docClient = DynamoDBDocumentClient.from(client, {
  marshallOptions: {
    removeUndefinedValues: true,
    convertEmptyValues: false,
  },
  unmarshallOptions: {
    wrapNumbers: false,
  },
});

export function getMainTable(): string {
  const table = process.env.MAIN_TABLE;
  if (!table) {
    throw new Error('MAIN_TABLE environment variable is not set');
  }
  return table;
}

export function getActivityTable(): string {
  const table = process.env.ACTIVITY_TABLE;
  if (!table) {
    throw new Error('ACTIVITY_TABLE environment variable is not set');
  }
  return table;
}

export function getLeadsTable(): string {
  const table = process.env.LEADS_TABLE;
  if (!table) {
    throw new Error('LEADS_TABLE environment variable is not set');
  }
  return table;
}

/** Convenience constants - read table names from env once at cold start */
export const MAIN_TABLE = process.env.MAIN_TABLE || '';
export const ACTIVITY_TABLE = process.env.ACTIVITY_TABLE || '';
export const LEADS_TABLE = process.env.LEADS_TABLE || '';

File: backend/layers/common/python/dynamo_client.py
import os
import boto3
from boto3.dynamodb.conditions import Key, Attr

# Table name constants from environment
MAIN_TABLE = os.environ.get('MAIN_TABLE', 'CloudAIP-Main-staging')
ACTIVITY_TABLE = os.environ.get('ACTIVITY_TABLE', 'CloudAIP-Activity-staging')
LEADS_TABLE = os.environ.get('LEADS_TABLE', 'CloudAIP-Leads-staging')

# DynamoDB resource (reused across invocations)
dynamodb = boto3.resource('dynamodb')

def get_main_table():
    return dynamodb.Table(MAIN_TABLE)

def get_activity_table():
    return dynamodb.Table(ACTIVITY_TABLE)

def get_leads_table():
    return dynamodb.Table(LEADS_TABLE)
